### DISCUSSION TITLE
.upload() is not a function?

### DISCUSSION THREAD

**Discussion by AliAryanTech** [qgustavor/mega#137] [Q&A] [answered]
**Describe the bug**
Well, I'm trying to test .upload but the terminal saying it's not a function? even I tried all examples of .upload it's saying it's not a function. 
also storage.once(), saying the same it's not a function.

**To Reproduce**
```js
const { Storage } = require('megajs')

const storage = new Storage({
  email: '****', // email address 
  password: '*****' // my password 
}, error => {
  return console.log(`${error ?  "Some error happened" : "User is now logged in"}`)
}).ready 

storage.upload('hello-world.txt', 'Hello world!', (error, file) => {
  if (error) return console.error('There was an error:', error)
  console.log('The file was uploaded!', file)
})
```

**Expected behavior**
I tried in termux & replit it test the code!
Termux
![termux screenshot](https://user-images.githubusercontent.com/85696485/209958929-e4c9836c-d0db-4423-abb8-bb785b10a5dc.jpg)

Replit:
![replit screenshot](https://user-images.githubusercontent.com/85696485/209958741-e0778bed-36d6-47cf-8974-dad668f8004b.jpg)

**Additional context**
One more thing can you tell me an example with a Buffer? Like how to upload a Buffer, I saw those example but coz of upload function unable to test upload by buffer, that's why I thought it would be better if you tell me an example!!

```js
let buffer = await getBuffer("https://telegra.ph/file/e991bb4b535a0f1425aa0.jpg")

const file = await storage.upload('hello-world.txt', Buffer.from(buffer, 'base64')).complete
console.log('The file was uploaded!', file)
```
Like this? That's just a buffer, I don't know what's the file name and size!?
Then why this, ```allowUploadBuffering: true```

**Comment by qgustavor** [comment, marked as answer]
I'm quite confused on what you're trying to do, I don't know your JavaScript background, but seems you are mistaking how asynchronous JavaScript works.

Do you know how to use Promises? And how about callbacks? If not check some tutorial for those topics first.

`storage.ready` is a Promise, so seems you're missing an `await` keyword, but at the same time you defined a callback at the same time, which… is weird. Use one or another. Callbacks is the old pattern, promises is the new one, this library support both because backyards compatibility but I recommend learning and using promises as they avoid some issues that callbacks have (like nesting hell).

Seems you are using Node CJS, it does not support top level await, check examples in mega.js.org to workaround this limitation (use a async IIFE). If possible learn and use ESM, it does not have this limitation.

Try learning about async patterns in JavaScript, them fix the login first, that's a lot for now, I will answer the upload question later.

~~I'm on the mobile client, when I get home I will move this into a discussion because that's not an issue with the library, it's mostly a JavaScript related question.~~ **Edit:** issue migrated to a discussion.

**Reply by AliAryanTech** [reply]
Indeed, I'm using Nodejs
But my login working fine with that code. 🤔 now I see I was consoling file, that's why!!

```node
const a = async () => {
const { Storage } = require('megajs')

const storage = await new Storage({
  email: '****',
  password: '****'
}, error => {
  return console.log(`${error ?  "Some error happened" : "User is now logged in"}`)
}).ready 

const file = await storage.upload('hello-world.txt', 'Hello world!').complete
console.log('The file was uploaded!')
}
a()
```
Here:
![screenshot](https://user-images.githubusercontent.com/85696485/209995215-cf5d2f65-d537-428a-bc14-83136b996437.jpg)

I thought it's error, when console file: 
```node
console.log('The file was uploaded!', file)
```
Here:
![screenshot](https://user-images.githubusercontent.com/85696485/209995753-f3fc6bb1-62dd-4bc8-85dd-803161648a9e.jpg)

**About Buffer upload:**
```node
const buffer = await getBuffer("https://telegra.ph/file/e991bb4b535a0f1425aa0.jpg")

const file = await storage.upload('Buffer.png', Buffer.from(buffer, 'base64')).complete
console.log('The file was uploaded!')
```
So, it works fine:
![screenshot](https://user-images.githubusercontent.com/85696485/209996459-5f5ec7fe-56c7-4b9d-aea8-04448b411952.jpg)

Here:
![screenshot](https://user-images.githubusercontent.com/85696485/209996355-dee7c9a1-e4dc-477b-b993-36eabeaee166.jpg)

Well, Is there any other way to upload Buffer!?
when I only have buffer don't know size, mimetype, name. Just asking 🤔

**Reply by qgustavor** [reply]
> when I only have buffer don't know size, mimetype, name. Just asking 

MEGA requires a file name, you can use any name you want, sometimes I just use "a" just to fill this requirement when I don't care about the file name (which is rare, most people don't want to download something without a name or with a meaningless name).

Mime type is not needed. You need it for HTTP servers, but not for this library, MEGA does not check it at all. If you have a Buffer then its size is just `buffer.length`, if you pass a buffer to the library [it will just get the size from this property](https://github.com/qgustavor/mega/blob/ee088b4b21500cf3d41b909419d8230d25c28305/lib/mutable-file.mjs#L253-L257) and upload the buffer, you don't even need to worry about this.

By the way, your `getBuffer` function returns a Base64 string? Because you are taking the result from it and converting from Base64 by using `Buffer.from(buffer, 'base64')`. If the `getBuffer` function returns a Buffer (which is what its name says) then you can just pass it to `upload`. If it returns a ArrayBuffer or a Uint8Array (like if this function this calls `url => fetch(url).then(resp => resp.arrayBuffer()`) you can pass those too to `upload` if you wish.

**Reply by AliAryanTech** [reply]
Indeed, ```getBuffer``` function return 'arraybuffer'
As you said I can simply pass it:
```node
const file = await storage.upload('Buffer.png', buffer).complete
```
**it's working fine, thank you!**

**Comment by AliAryanTech** [comment]
By the way, how to upload file in new folder ```storage.mkdir()``` and whenever we upload any file it only upload to that specific folder, I don't want it create one more new folder of the same name.

Can you give an example?

**Reply by qgustavor** [reply]
```javascript
import { Storage } from 'megajs'
const storage = await new Storage({/* credentials */}).ready
const newFolder = await storage.mkdir('name of the folder')
const fileInNewFolder = await newFolder.upload('name of the file.txt', 'contents of the file')
```

**Reply by AliAryanTech** [reply]
> ```js
> import { Storage } from 'megajs'
> const storage = await new Storage({/* credentials */}).ready
> const newFolder = await storage.mkdir('name of the folder')
> const fileInNewFolder = await newFolder.upload('name of the file.txt', 'contents of the file')
> ```

Well, I asked for a specific folder, 

when I run the code twice times:
![screenshot](https://user-images.githubusercontent.com/85696485/210137310-5483992d-06b3-43a7-8788-b87a3beca637.jpg)

Then this happened:
 ![screenshot](https://user-images.githubusercontent.com/85696485/210137302-03389c10-c60c-46ff-b0ca-5d2cae220efc.jpg)

**Reply by qgustavor** [reply]
So you want to reuse an existent folder instead of making one and uploading:

```javascript
// Import library
import { Storage } from 'megajs'
// Login to account
const storage = await new Storage({/* credentials */}).ready
// Get folder from account
const folderName = 'name of the folder'
let folder = storage.root.children.find(e => e.name === folderName && e.directory)
// If no folder was found, create a new one
if (!folder) folder  = await storage.mkdir(folderName)
// Upload file to folder
const fileInNewFolder = await folder.upload('name of the file.txt', 'contents of the file')
```

Look, [documentation is here](https://mega.js.org/docs/api). If it is too hard to understand I'm thinking on adding an "Examples" section in the documentation, I just need inspiration about what kind of examples to add.

**Reply by AliAryanTech** [reply]
**Thank you for your help & for your precious time!**

**Reply by AliAryanTech** [reply]
Sorry, for disturbing but can you tell me?
I got this error while uploading buffer, is this because the file size big or something?

```node
const file = await storage.upload('Buffer.mp4', buffer).complete
```
Response:
```Specify a file size or set allowUploadBuffering to true```

**Reply by qgustavor** [reply]
Check what `buffer` is. Maybe that's `undefined` or a Promise (maybe you forgot an `await`, it happened with me a lot).

**Reply by AliAryanTech** [reply]
> Check what `buffer` is. Maybe that's `undefined` or a Promise (maybe you forgot an `await`, it happened with me a lot).

Nope, I mostly use `await` that file is WhatsApp.webp, I tested it on video, photo, audio and documents. Only that webp File give that error.
One more thing, if mine file size big then I can simply use that code right?
```node
const file = await storage.upload('Buffer.mp4', buffer).complete
```

**Reply by qgustavor** [reply]
You can use that code as long you can hold the entire file in memory, otherwise use streaming: https://mega.js.org/docs/tutorial/uploading

```javascript
const file = await storage.upload({
  name: 'hello-world.txt',
  size: 12
}, fs.createReadStream('hello-world.txt')).complete
console.log('The file was uploaded!', file)
```

**Reply by AliAryanTech** [reply]
Instead reply, Wow
Ok, got thank you!

**Reply by AliAryanTech** [reply]
It's just an idea while using `.download()` user don't know what's the mimetype of the buffer file. So, it would be good if you add mimetype in file.loadAttributes() in your function, with next version!!

I know, I can do this;
```js
import {fileTypeFromBuffer} from 'file-type';

console.log(await fileTypeFromBuffer(buffer));
//=> {ext: 'png', mime: 'image/png'}
```

**Reply by qgustavor** [reply]
I don't understand, why? The library does not need to handle mime types, MEGA does not require it so it's out of scope. If you want to detect mime types there are other libraries for that, I remember using one on direct-mega.

**Reply by AliAryanTech** [reply]
I need ext or mimetype, for now I'm using `file-type`

 It's just while using `.upload()`
If user upload his files, It would be weird to upload a file without extension.
```node
const file = await storage.upload('Buffer', buffer).complete
```

`.download()`, Also I'm sending response in different media, like 'audio', 'video', 'document', etc.
That's why, I thought it would be good if `file.loadAttributes()`, also send ext, and mime.

**Reply by qgustavor** [reply]
That's out of scope. Use a library specific for that, like the one you cited.

**Reply by AliAryanTech** [reply]
> That's out of scope. Use a library specific for that, like the one you cited.

I was just trying to give an idea, **nevermind Ok, and thank you!**

**Reply by AliAryanTech** [reply]
Well, Is this possible, to see file downloading progress in console? Sometimes, it took pretty long while downloading...

and sometimes it get `Killed` and program exit.

**Reply by qgustavor** [reply]
https://www.npmjs.com/package/progress-stream https://www.npmjs.com/package/cli-progress

**Reply by AliAryanTech** [reply]
> https://www.npmjs.com/package/progress-stream https://www.npmjs.com/package/cli-progress

I see, btw why the downloading take too much time when the file is nearly +100mb. It should be fast coz, program running on hosting server not localhost?

Well, if you familiar with `progress-stream` module can you give me example with your `megajs`
